### PR TITLE
GitHub auth

### PR DESCRIPTION
**Pull request by its-clem-fandango**

### CODE REVIEW

File: apps/client/app/dashboard/dashboard.tsx [hunk 1]
```diff
@@ -28,4 +55,53 @@ function Dashboard() {
...
+
+export const getServerSideProps: GetServerSideProps<DashboardProps> = async (
+  context,
+) => {
+  // Retrieve the session Id from cookies attached to incoming request
+  const token = context.req.cookies.sessionId
+
+  try {
+    // Send a request to backend's session validation endpooint with the sessionId so server can use  sessionId to look up session and determine if its valid
+    const validationResponse = await fetch(
```
Comment (craigostrin) [inline apps/client/app/dashboard/dashboard.tsx:69]: I would modularize this

**Review by craigostrin — CHANGES_REQUESTED** [review]

File: apps/client/app/dashboard/dashboard.tsx [hunk 2]
```diff
@@ -4,10 +4,39 @@ import { columns } from "@/app/dataTable/columns"
...
 
+  console.log("Dashboard: Checking user state", user)
+
+  useEffect(() => {
+    console.log("Dashboard: useEffect, user:", user)
+    if (isLoggedIn === false) {
+      console.log("Dashboard: User not logged in, redirecting to /login")
+      // router.push("/login")
+    }
+  }, [isLoggedIn])
```
Comment (Imagine96) [inline apps/client/app/dashboard/dashboard.tsx:33]: what is the point of this? if its just logs we should clean them.

File: apps/client/app/dashboard/dashboard.tsx [hunk 3]
```diff
@@ -4,10 +4,39 @@ import { columns } from "@/app/dataTable/columns"
...
+
+  /*   if (!user) {
+    console.log("Dashboard: Rendering null due to no user")
+    return null
+  } */
+  console.log(races)
```
Comment (Imagine96) [inline apps/client/app/dashboard/dashboard.tsx:39]: remove logs

File: apps/client/helper/apicalls.tsx [hunk 4]
```diff
@@ -11,4 +12,24 @@ async function getChallangeData(id: number) {
...
+      Cookie: `sessionId=${sessionId}`,
+    },
+  }
+
+  const response = await fetch(`${API_URL}/session/validateSession`, options)
+  if (!response.ok) {
+    throw new Error("Session validation failed")
+  }
+  const userDetails = await response.json()
+  console.log("User from dashboard returns successfully: ", userDetails)
```
Comment (Imagine96) [inline apps/client/helper/apicalls.tsx:31]: remove logs

File: apps/client/lib/useAuth.tsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    const value = { user, isLoggedIn, setUser }
+  
+    return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
+}
+
+  
+  // Custom hook to use authentication context
+  export const useAuth = () => {
+    const context = useContext(AuthContext)
+    console.log("Current user in useAuth:", context.user)
```
Comment (Imagine96) [inline apps/client/lib/useAuth.tsx:67]: remove logs

File: apps/server/src/auth/auth.controller.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    const tokenResponse = await this.exchangeCodeForToken(
+      code,
+      clientID,
+      clientSecret,
+    );
+    const githubUser = await this.fetchGithubUserProfile(
+      tokenResponse.access_token,
+    );
+
+    console.log("****GITHUB USER OBJECT******", githubUser);
```
Comment (Imagine96) [inline apps/server/src/auth/auth.controller.ts:33]: remove logs

File: apps/server/src/auth/auth.controller.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    res.cookie("sessionId", session._id.toString(), {
+      httpOnly: true,
+      secure: true,
+      sameSite: "strict",
+      maxAge: 6666666,
+      path: "/",
+    });
+
+    //REPLACE WITH /dashboard ENDPOINT
+    console.log(`TOKEN RESPONSE: ${JSON.stringify(tokenResponse)}`);
```
Comment (Imagine96) [inline apps/server/src/auth/auth.controller.ts:56]: remove logs

File: apps/server/src/auth/auth.controller.ts [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        path: "/login/oauth/access_token",
+        method: "POST",
+        headers: {
+          "Content-Type": "application/x-www-form-urlencoded",
+          Accept: "application/json", // Requests that GitHub responds with JSON
+        },
+      };
+
+      // Making the HTTPS POST request to GitHub's token endpoint.
+      const req = https.request(options, (res) => {
```
Comment (Imagine96) [inline apps/server/src/auth/auth.controller.ts:89]: why a "vanilla" http req here? why not use fetch as in the rest of the project?

File: apps/server/src/users/users.service.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+    const user = await this.userModel.findOneAndUpdate(
+      { username },
+      { username, avatarURL: avatarURL },
+      {
+        new: true,
+        upsert: true,
+      },
+    );
+    console.log("USER OBJECT WITH AVATAR URL: ", user.toObject());
```
Comment (Imagine96) [inline apps/server/src/users/users.service.ts:25]: logs

**Review by Imagine96 — CHANGES_REQUESTED** [review]
Things seems alright, just had a couple questions and there are a LOT of console.logs. I don't see a reason why it would fail if the nest and mongoose syntax is good, which it looks like. The thing that called my attention is the use of the direct http package in some places instead of fetch, is there a reason for that?

File: apps/server/src/auth/auth.controller.ts (hunk 6 above)
Comment (its-clem-fandango) [reply apps/server/src/auth/auth.controller.ts:33]: keeping this one just incase i need some of the data

**Review by Imagine96 — APPROVED** [review]

**Review by craigostrin — APPROVED** [review]

**Pull request merged**